### DISCUSSION TITLE
Interest in multi schema zod resolver?

### DISCUSSION THREAD

**Discussion by brmenchl** [react-hook-form/react-hook-form#4815] [Ideas]
I had a use case where a single field in a form needed to change the validation schema of the entire form (determining whether the model was "Published" or not. Published models had more strict validation).

I ended up creating a custom zod resolver that reused almost everything from the existing zodResolver, but allowed you to pass in a function `(values: FieldValues) => z.ZodSchema<any, any>` that you can use to swap out zod schemas at run time.

I could add this as a new zod resolver, but I wanted to hear what people think about it.

**Comment by jorisre** [comment]
@brmenchl can you share some code please ?

**Reply by brmenchl** [reply]
sure!

```typescript
type MultiSchemaResolver = <T, P extends UnpackNestedValue<T>>(
  schemaGetter: (values: P) => z.ZodSchema<any, any>
) => (
  values: P,
  context: any,
  options: ResolverOptions<T>
) => Promise<ResolverResult<T>>;

export const multiSchemaZodResolver: MultiSchemaResolver = (
  schemaGetter
) => async (values, _, options) => {
  try {
    return {
      errors: {},
      values: await schemaGetter(values).parseAsync(values),
    };
  } catch (error) {
    return {
      values: {},
      errors: error.isEmpty
        ? {}
        : toNestError(parseErrorSchema(error.errors, false), options.fields),
    };
  }
};
```

**Reply by brmenchl** [reply]
so, specific to my own use case i removed some of the arguments that the original `zodResolver` had, but those can definitely be added back. The body of the resolver is the same except for the line
`values: await schemaGetter(values).parseAsync(values),`, instead of just using one passed in `schema`

**Reply by brmenchl** [reply]
I use it like
```typescript
const getSchema = (values: FormInput) =>
  values.status === 'Published'
    ? publishedSchema
    : nonPublishedSchema;

// ...

useForm<FormInput>({
    resolver: multiSchemaZodResolver(getSchema),
    ...
  });
```

**Reply by bluebill1049** [reply]
you may have to build your own custom resolver, we current only support single schema per form, however you can use context to switch different schema as well.

**Reply by whengely** [reply]
I am running into this same issue.  What methods would I look at to use context to switch schema?  I'm not finding anything searching zod or rhf's docs.  Thanks

**Reply by bluebill1049** [reply]
https://react-hook-form.com/api/useform

<img width="957" alt="Screen Shot 2021-04-29 at 8 50 47 am" src="https://user-images.githubusercontent.com/10513364/116482189-0114d480-a8c8-11eb-9a0b-0c3648a5e5b8.png">

**Reply by ashrafnazar** [reply]
@bluebill1049 - thanks for this, but the example provided is not that intuitive. Could you possibly make the example a little more detailed?
